src/cluster_legacy.c

```diff
@@ -2990,7 +2991,51 @@ int clusterProcessPacket(clusterLink *link) {
 
                 if (clusterNodeIsMaster(sender)) {
                     /* Master turned into a slave! Reconfigure the node. */
-                    clusterDelNodeSlots(sender);
+                    if (master && !memcmp(master->shard_id, sender->shard_id, CLUSTER_NAMELEN)) {
+                        /* 'sender' was a primary and was in the same shard as `master`, its new primary */
+                        if (sender->configEpoch > senderConfigEpoch) {
+                            serverLog(LL_WARNING,
+                                    "Ignore stale gossip message from %.40s (%s) in shard %.40s;"
+                                    " gossip config epoch: %llu, current config epoch: %llu", 
+                                    sender->name,
+                                    sender->human_nodename,
+                                    sender->shard_id,
+                                    (unsigned long long)senderConfigEpoch,
+                                    (unsigned long long)sender->configEpoch);
+                        } else {
+                            /* A failover occurred in the shard where `sender` belongs to and `sender` is no longer
+                             * a primary. Update slot assignment to `master`, which is the new primary in the shard */
+                            int slots = clusterMoveNodeSlots(sender, master);
+                            serverLog(LL_NOTICE, "A failover occurred in shard %.40s; node %.40s (%s)"
+                                    " lost %d slot(s) to node %.40s (%s)",
+                                    sender->shard_id,
+                                    sender->name,
+                                    sender->human_nodename,
+                                    slots,
+                                    master->name,
+                                    master->human_nodename);
+                        }
+                    } else {
+                        int slots = clusterDelNodeSlots(sender);
+                        if (master != NULL) {
+                            /* `sender` was moved to another shard and has become a replica, remove its slot assignment */
+                            serverLog(LL_NOTICE, "Node %.40s (%s) moved from shard %.40s to shard %.40s"
+                                    " and became a slave; removed all %d slot(s) it used to own",
+                                    sender->name,
+                                    sender->human_nodename,
+                                    sender->shard_id,
+                                    master->shard_id,
+                                    slots);
+                        } else {
+                            /* We have no knowledge of `master` */
+                            serverLog(LL_NOTICE, "Node %.40s (%s) is not longer a master in shard %.40s;"
```

Comment:
Unless you feel strongly, I think we really only need this log message for both cases in the else case. Most of this code is just a lot of detailed log messages, and the logging hurts readability of this code.

Reply:
but it helps with debugging/troubleshooting, right? :-)

I have had many times wished that there are more logging statements when chasing down unexpected topology states. It is always hard (if even possible) to reproduce a consistency issue in a distributed system so logs become the best debugging tool. So unless we believe the logging would have impact on the performance (which shouldn't in this case), I am more in favor of keeping them.

Reply:
My concern wasn't performance, more so that it looked like there was 4 different edge cases being handled, when in reality there was just 2 but we have 4 if/else just to log a little bit more information. We could perhaps even compact it to 1 Just by doing a null check on the shard ID. Let me make a suggestion with what I mean just so we're on the same page (this isn't a blocker and I'll give off the cluster tests in either case in preparation for merging).